src/Build/BuildCheck/Analyzers/DoubleWritesAnalyzer.cs

```diff
@@ -0,0 +1,126 @@
+    }
+
+    /// <summary>
+    /// Contains the first project file + task that wrote the given file during the build.
+    /// </summary>
+    private readonly Dictionary<string, (string projectFilePath, string taskName)> _filesWritten = new(StringComparer.CurrentCultureIgnoreCase);
+
+    private void TaskInvocationAction(BuildCheckDataContext<TaskInvocationAnalysisData> context)
+    {
+        // This analyzer uses a hard-coded list of tasks known to write files.
+        switch (context.Data.TaskName)
+        {
+            case "Csc":
+            case "Vbc":
+            case "Fsc": AnalyzeCompilerTask(context); break;
+            case "Copy": AnalyzeCopyTask(context); break;
+        }
+    }
+
+    private void AnalyzeCompilerTask(BuildCheckDataContext<TaskInvocationAnalysisData> context)
+    {
+        var taskParameters = context.Data.Parameters;
+
+        // Compiler tasks have several parameters representing files being written.
+        AnalyzeParameter("OutputAssembly");
+        AnalyzeParameter("OutputRefAssembly");
+        AnalyzeParameter("DocumentationFile");
+        AnalyzeParameter("PdbFile");
+
+        void AnalyzeParameter(string parameterName)
+        {
+            if (taskParameters.TryGetValue(parameterName, out TaskParameter? taskParameter))
+            {
+                string outputPath = taskParameter.EnumerateStringValues().FirstOrDefault() ?? "";
+                AnalyzeWrite(context, outputPath);
+            }
+        }
+    }
+
+    private void AnalyzeCopyTask(BuildCheckDataContext<TaskInvocationAnalysisData> context)
+    {
+        var taskParameters = context.Data.Parameters;
+
+        // The destination is specified as either DestinationFolder or DestinationFiles.
+        if (taskParameters.TryGetValue("SourceFiles", out TaskParameter? sourceFiles) &&
+            taskParameters.TryGetValue("DestinationFolder", out TaskParameter? destinationFolder))
+        {
+            string destinationFolderPath = destinationFolder.EnumerateStringValues().FirstOrDefault() ?? "";
+            foreach (string sourceFilePath in sourceFiles.EnumerateStringValues())
+            {
+                AnalyzeWrite(context, Path.Combine(destinationFolderPath, Path.GetFileName(sourceFilePath)));
+            }
+        }
+        else if (taskParameters.TryGetValue("DestinationFiles", out TaskParameter? destinationFiles))
+        {
+            foreach (string destinationFilePath in destinationFiles.EnumerateStringValues())
+            {
+                AnalyzeWrite(context, destinationFilePath);
+            }
+        }
+    }
+
+    private void AnalyzeWrite(BuildCheckDataContext<TaskInvocationAnalysisData> context, string fileBeingWritten)
+    {
+        if (!string.IsNullOrEmpty(fileBeingWritten))
+        {
+            // Absolutize the path.
+            fileBeingWritten = Path.GetFullPath(fileBeingWritten, context.Data.ProjectFilePath);
```

Comment:
Can actually relative paths appear here? If yes - should they actually be absolutized by appending to the dir location of the project being analyzed (which should be part of the data context)?

Reply:
Relative paths can appear here. And because the build runs with Current Working Directory == the-directory-of-the-project-being-built, effectively they are relative to the entry point project file. So I think this was _almost_ correct, it was just mistakenly using the path of the project file instead of just the directory.

Reply:
The test I'm using for this looks like this:

entry.proj:
```xml
<Project>
  <Import Project="subdir\proj.proj" />
</Project>
```

subdir\proj.proj:
```xml
<Project>
  <Target Name="Build">
    <Copy SourceFiles="src.file" DestinationFiles="dst.file" />
    <Copy SourceFiles="src.file" DestinationFiles="dst.file" />
    <Message Text="Hello world" />
  </Target>
</Project>
```

Even though `src.file` and `dst.file` appear in a project file imported from a subdirectory, they refer to files alongside `entry.proj` when `entry.proj` is built.

Reply:
Oh - you are right - I was misunderstanding your code - it was perfectly correct. And now it is even more correct :-))
Thank you!